qiskit_optimization/algorithms/qrao/magic_rounding.py

```diff
@@ -0,0 +1,448 @@
+
+        Returns:
+            MagicRoundingResult: The results of the magic rounding process.
+
+        Raises:
+            NotImplementedError: If the circuit is not available for magic rounding.
+            ValueError: If the sampler is not configured with a number of shots.
+        """
+        expectation_values = rounding_context.expectation_values
+        circuit = rounding_context.circuit
+        q2vars = rounding_context.encoding.q2vars
+        var2op = rounding_context.encoding.var2op
+        vars_per_qubit = rounding_context.encoding.max_vars_per_qubit
+
+        if circuit is None:
+            raise NotImplementedError(
+                "Magic rounding requires a circuit to be available.  Perhaps try "
+                "semideterministic rounding instead."
+            )
+
+        if self._sampler.options.get("shots") is None:
+            raise ValueError(
+                "Magic rounding requires the sampler to be configured with a number of shots."
+            )
+        self._shots = self._sampler.options.shots
+
+        if self.basis_sampling == "uniform":
+            # uniform sampling
+            bases, basis_shots = self._sample_bases_uniform(q2vars, vars_per_qubit)
+        else:
+            # weighted sampling
+            if expectation_values is None:
+                raise NotImplementedError(
+                    "Magic rounding with weighted sampling requires the expectation values of the "
+                    "``RoundingContext`` to be available, but they are not."
+                )
+            bases, basis_shots = self._sample_bases_weighted(
+                q2vars, expectation_values, vars_per_qubit
+            )
+
+        # For each of the Magic Bases sampled above, measure
+        # the appropriate number of times (given by basis_shots)
+        # and return the circuit results
+        basis_counts = self._evaluate_magic_bases(circuit, bases, basis_shots, vars_per_qubit)
+        # keys will be configurations of decision variables
+        # values will be total number of observations.
+        soln_counts = self._compute_dv_counts(basis_counts, bases, var2op, vars_per_qubit)
+
+        soln_samples = [
+            SolutionSample(
+                x=np.asarray([int(bit) for bit in soln]),
+                fval=rounding_context.encoding.problem.objective.evaluate(
+                    [int(bit) for bit in soln]
+                ),
+                probability=count / self._shots,
+                status=OptimizationResultStatus.SUCCESS
+                if rounding_context.encoding.problem.is_feasible([int(bit) for bit in soln])
+                else OptimizationResultStatus.INFEASIBLE,
+            )
+            for soln, count in soln_counts.items()
+        ]
+
+        assert np.isclose(
+            sum(soln_counts.values()), self._shots
+        ), f"{sum(soln_counts.values())} != {self._shots}"
+        assert len(bases) == len(basis_shots) == len(basis_counts)
```

Comment:
Wondering about the assert here. Should we have a runtime check here - is this something a user may run into.

Reply:
No, it's really just a check on the consistency of the above code.  If it was something I expected the user to run in to, I would have raised an exception instead.  Or, in other words, since it is possible to disable asserts by passing `-O` to the interpreter, it should only be used to test the internal consistency of a module, and that is how it is used here.